### PR TITLE
Random lowering bugfixes

### PR DESCRIPTION
**Pull request by mlechu**
Some bugs I ran into while hooking up legacy macros and running random test files from base julia

- `init` not provided for possibly-empty `sum()`s in desugaring causing failures for empty ctors, empty tuple destruct
- linearize: break_block type error
- missing desugaring when if-condition is a block
- boundserror in `add_ir_debug_info!`
- undef var `ex` in `expand_arrow_arglist`

### CODE REVIEW

File: src/eval.jl [hunk 1]
```diff
@@ -84,7 +84,7 @@ function add_ir_debug_info!(current_codelocs_stack, stmt)
                                   current_codelocs_stack[j][1] != locstk[j][1])
             while length(current_codelocs_stack) >= j
                 info = pop!(current_codelocs_stack)
-                push!(last(current_codelocs_stack)[2], info)
+                !isempty(current_codelocs_stack) && push!(last(current_codelocs_stack)[2], info)
```
Comment (c42f) [inline src/eval.jl:87]: What was the example code which triggered this?

**Review by c42f — COMMENTED** [review]
Thanks, these fixes look good.

Please add tests for each of the syntax edge cases hit here :)

File: src/eval.jl (hunk 1 above)
Comment (mlechu) [reply src/eval.jl:87]: Can't remember unfortunately, so I'm not sure it's the right fix and I'll undo this for now.  I think it was something to do with the filename violating assumptions and/or very short IR.

File: src/desugaring.jl [hunk 2]
```diff
@@ -2124,7 +2124,7 @@ function expand_decls(ctx, ex, is_const=false)
             @chk numchildren(binding) == 2
             lhs = strip_decls!(ctx, stmts, declkind, declmeta, binding[1])
             push!(stmts, expand_assignment(ctx, @ast ctx binding [kb lhs binding[2]]))
-        elseif is_sym_decl(binding) && !is_const
+        elseif is_sym_decl(binding)
```
Comment (aviatesk) [inline src/desugaring.jl:2127]: Does this fix https://github.com/c42f/JuliaLowering.jl/issues/28 too?

File: src/eval.jl (hunk 1 above)
Comment (c42f) [reply src/eval.jl]: Ok!  I ask because it's not super clear to me that this is the correct fix (constructing the correct debug info format is pretty confusing)

File: test/destructuring.jl [hunk 3]
```diff
@@ -36,6 +36,17 @@ let
...
+@test_throws JuliaLowering.LoweringError JuliaLowering.include_string(test_mod, """
+let
+    (x,) = ()
+end
+""") == ([1, 2, 3], 4)
+@test_throws JuliaLowering.LoweringError JuliaLowering.include_string(test_mod, """
+let
+    (x,y,ys...) = (1,)
+end
+""")
```
Comment (c42f) [inline test/destructuring.jl:48]: Looks good, let's move both of these test cases into destructuring_ir.jl.  The IR tests have good infrastructure for pretty printing errors which make such tests nice to read and write. For example:

https://github.com/c42f/JuliaLowering.jl/blob/cfa6901d597dc27a303c28cc87fed608eb76df2a/test/destructuring_ir.jl#L85

File: test/loops.jl [hunk 4]
```diff
@@ -95,6 +95,18 @@ let
     a
 end
 """) == [1, 2]
+@test JuliaLowering.include_string(test_mod, """
```
Comment (c42f) [inline test/loops.jl:98]:
```suggestion
# Break from inner nested loop
@test JuliaLowering.include_string(test_mod, """
```

**Review by c42f — APPROVED** [review]
Looks good. Let's move a couple of the error tests into IR tests.
